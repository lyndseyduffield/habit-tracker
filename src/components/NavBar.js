import React from "react";
import { Link, withRouter } from "react-router-dom";

class NavBar extends React.Component {
  state = {
    collapsed: false
  };

  toggleView = event => {
    event.preventDefault();
    this.setState(prevState => ({
      collapsed: !prevState.collapsed
    }));
  };

  changeViewButton = () => {
    if (this.state.collapsed) {
      return <strong>Expand View</strong>;
    } else {
      return <strong>Collapse View</strong>;
    }
  };

  renderDisplay() {
    if (this.props.location.pathname === "/") {
      return (
        <nav class="navbar has-shadow">
          <div class="navbar-brand">
            <div class="navbar-item">
              <strong>HabitTracker</strong>
            </div>
          </div>
          <div class="navbar-end">
            <div class="navbar-item">
              <div class="buttons">
                <Link to="/new" class="button is-light">
                  <strong>New Habit</strong>
                </Link>
                <div
                  onClick={event => this.toggleView(event)}
                  class="button is-light"
                >
                  {this.changeViewButton()}
                </div>
                <Link to="/" class="button is-light">
                  Home
                </Link>
              </div>
            </div>
          </div>
        </nav>
      );
    } else {
      return (
        <nav class="navbar has-shadow">
          <div class="navbar-brand">
            <div class="navbar-item">
              <strong>HabitTracker</strong>
            </div>
          </div>
          <div class="navbar-end">
            <div class="navbar-item">
              <div class="buttons">
                <Link to="/new" class="button is-light">
                  <strong>New Habit</strong>
                </Link>
                <Link to="/" class="button is-light">
                  Home
                </Link>
              </div>
            </div>
          </div>
        </nav>
      );
    }
  }

  render() {
    return <div>{this.renderDisplay()}</div>;
  }
}

export default withRouter(NavBar);
