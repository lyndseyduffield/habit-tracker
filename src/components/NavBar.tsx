import React from "react";
import { Link, withRouter, RouteComponentProps } from "react-router-dom";
import { connect, ConnectedProps } from "react-redux";
import { updateCurrentUser } from "../store/actions";
import { getHabitIds } from "../utils";
import { State } from "../store/types";

const mapState = (state: State) => {
  const user = state.currentUser;
  if (user) {
    return {
      habitIds: getHabitIds(state.userStates[user].habits),
    };
  } else {
    return {
      habitIds: [],
    };
  }
};

const connector = connect(mapState);

type PropsFromRedux = ConnectedProps<typeof connector>;

type OwnProps = {
  collapsed: boolean;
  toggleCollapse: (event: React.MouseEvent) => void;
};

type Props = PropsFromRedux & RouteComponentProps & OwnProps;

class NavBar extends React.Component<Props> {
  buttonLabel = () => {
    if (this.props.collapsed) {
      return <strong>Expand View</strong>;
    } else {
      return <strong>Collapse View</strong>;
    }
  };

  handleLogout = () => {
    let username = null;
    this.props.dispatch(updateCurrentUser(username));
    this.props.history.push("/");
  };

  render = () => {
    const showView =
      this.props.location.pathname === "/home" &&
      this.props.habitIds.length > 0;
    return (
      <nav className="navbar has-shadow">
        <div className="navbar-brand">
          <Link to="/" className="navbar-item">
            <strong>HabitTracker</strong>
          </Link>
        </div>
        <div className="navbar-end">
          <div className="navbar-item">
            <div className="buttons">
              <Link to="/new" className="button is-light">
                <strong>New Habit</strong>
              </Link>
              {showView && (
                <div
                  onClick={(event) => this.props.toggleCollapse(event)}
                  className="button is-light"
                >
                  {this.buttonLabel()}
                </div>
              )}
              <button
                onClick={() => this.handleLogout()}
                className="button is-light"
              >
                <strong>Logout</strong>
              </button>
            </div>
          </div>
        </div>
      </nav>
    );
  };
}

const ConnectedNavBar = connector(NavBar);

export default withRouter(ConnectedNavBar);
