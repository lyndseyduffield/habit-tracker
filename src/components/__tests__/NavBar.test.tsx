import React from "react";
import NavBar from "../NavBar";
import renderer from "react-test-renderer";
import { StaticRouter } from "react-router";
import { testReducer } from "../../reducers/__tests__/index.test";
import { createStore } from "redux";
import { Provider } from "react-redux";

const store = createStore(testReducer);

it("renders the navbar correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <StaticRouter>
          <NavBar collapsed={true} toggleCollapse={() => {}} />
        </StaticRouter>
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
