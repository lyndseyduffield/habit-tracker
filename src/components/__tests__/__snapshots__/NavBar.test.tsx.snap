// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the navbar correctly 1`] = `
<nav
  className="navbar has-shadow"
>
  <div
    className="navbar-brand"
  >
    <a
      className="navbar-item"
      href="/"
      onClick={[Function]}
    >
      <strong>
        HabitTracker
      </strong>
    </a>
  </div>
  <div
    className="navbar-end"
  >
    <div
      className="navbar-item"
    >
      <div
        className="buttons"
      >
        <a
          className="button is-light"
          href="/new"
          onClick={[Function]}
        >
          <strong>
            New Habit
          </strong>
        </a>
        <button
          className="button is-light"
          onClick={[Function]}
        >
          <strong>
            Logout
          </strong>
        </button>
      </div>
    </div>
  </div>
</nav>
`;
